packages/vm/src/cache.rs

```diff
@@ -116,16 +116,17 @@ where
         backend: Backend<S, A, Q>,
         options: InstanceOptions,
     ) -> VmResult<Instance<S, A, Q>> {
+        let store = make_store_headless(Some(options.memory_limit));
         // Get module from memory cache
-        if let Some(module) = self.memory_cache.load(checksum)? {
+        if let Some(module) = self.memory_cache.load(checksum, &store)? {
             self.stats.hits_memory_cache += 1;
             let instance =
-                Instance::from_module(module, backend, options.gas_limit, options.print_debug)?;
+                Instance::from_module(&module, backend, options.gas_limit, options.print_debug)?;
```

Comment:
Huh? how does ownership work here/compile?

`module` will be dropped the end of this block, but we need to own it as long as `instance` exists. Confused by the reference here - unless `self.memory_cache.load` returns a ref of data stored in the cache

Reply:
Okay, I looked in the function and I guess you never a ref to the module, just us it to initialize. I found this line, but haven't dug into wasmer api to see if it uses the lifetimes of either of those two inputs. I guess not.

`let wasmer_instance = Box::from(WasmerInstance::new(module, &import_obj)`

Reply:
`Module`s, `Environment`s, `Memory`s and other things are cheaply clonable now, using [interior mutability](https://doc.rust-lang.org/reference/interior-mutability.html) such that multiple closes operate on the same underlying data.

Reply:
After seeing `Arc<Artifact>` in your other PR, this makes much more sense